### PR TITLE
Serverless based Lambda function for Labelling PRs based on CI & PR Review status

### PR DESCRIPTION
**Pull request by ChaiBapchya**
# Purpose
An automated way of labeling PRs based on CI status & PR Review state.
This serves dual purpose
- Automates the labeling
- Allows contributors to prioritize review of PRs.

# Testing
Tested it locally & on dev account.
Deployed this on Prod account.

# Reproduce
Once you put the redacted code & have AWSCLI + Serverless configured
```
./deploy_lambda.sh
```
Should work for deploying this lambda function on the account [dev/prod stage based on choice].

** Note **
Secrets have been redacted.

# Approach
## Details
Github Webhook -> Lambda 1 -> SQS -> Lambda 2
Github Webhook sends Status Events & triggers the lambda function "pr-status-labeler"
Lambda 2 in turn leverages Github API to add labels to PR [based on the following logic]
```
if WIP in title or PR is draft or CI failed:
    pr-work-in-progress
elif CI has not started yet or CI is in progress:
    pr-awaiting-testing
else: # CI passed checks
    if pr has at least one approval and no request changes:
        pr-awaiting-merge
    elif pr has no review or all reviews have been dismissed/re-requested:
        pr-awaiting-review
    else: # pr has a review that hasn't been dismissed yet no approval
        pr-awaiting-response
```

#### ~~Details~~
~~Runs once per day. Configured in the `serverless.yml` script.
It reuses credentials set during MXNet Bot [Jenkins & Github credentials].~~

#### ~~Results~~
~~Results in the successful run on Prod for apache/incubator-mxnet repo~~
```
~~Open PRs : 207~~
~~PRs labeled : 18~~
~~PRs already labeled : 38~~
~~PRs with unknown jobs : 2~~
~~PRs with status failed : 149~~
```

~~Lambda function labeled  18 PRs that had passed all the test. Detailed logs in the Cloudwatch Log Group [corresponding to the Lambda function] in the Prod CI AWS account]~~

### CODE REVIEW

File: services/lambda-label-successful-prs/CIBot.py
Comment (marcoabreu) [inline services/lambda-label-successful-prs/CIBot.py]: Don't query Jenkins but query the GitHub API. Every PR states which checks are required - even before they are run. That list is always up to date and immediately available.

File: services/lambda-label-successful-prs/CIBot.py
Comment (marcoabreu) [inline services/lambda-label-successful-prs/CIBot.py]: 11?

File: services/lambda-label-successful-prs/CIBot.py
Comment (marcoabreu) [inline services/lambda-label-successful-prs/CIBot.py]: Why check Jenkins at all? All the information is already available through the GitHub checks API. No need to query Jenkins

**Review by marcoabreu — CHANGES_REQUESTED** [review]
I'm not really sure about the objective of this, but I see quite a few issues: 

- you don't need to query Jenkins at all, GitHub has all the information
- once a PR went into a successful status, the label will not be redacted
- you should rather use an event driven approach where you subscribe to the GitHub webhook Event which is fired every single time a commit is made or check has been updated. Then have a label like "ci-passed", "unstable" and "ci-unsuccesful".

Right now I think that this implementation will cause rather more confusion.

File: services/lambda-label-successful-prs/CIBot.py
Comment (ChaiBapchya) [reply services/lambda-label-successful-prs/CIBot.py]: I had considered that option of not using Github API. I looked into the PyGithub API documentation.
For every object of type PullRequest, I coudln't find a method that returns the checks corresponding to that PR.

E.g. https://pygithub.readthedocs.io/en/latest/github_objects/PullRequest.html

Hence I had to rely on Jenkins API

Comment (ChaiBapchya) [reply services/lambda-label-successful-prs/CIBot.py]: https://github.com/PyGithub/PyGithub/issues/1063 is an open issue for integrating PyGithub with Checks API. I had created a similar issue yesterday: https://github.com/PyGithub/PyGithub/issues/1621

Comment (ChaiBapchya) [reply services/lambda-label-successful-prs/CIBot.py]: Option 1: Stick to Jenkins API
Option 2: Use Githubv3 REST API
I could use curl/requests to make REST API calls to GitHub [instead of Jenkins].

Option 3: Use github3 python package [https://github.com/sigmavirus24/github3.py]
+ Python-native calls via easy-to-use interface
+ Mentioned on Github Developer libraries: https://developer.github.com/v3/libraries/
- Not sure if its the standard. Less stars.

Before doing that, I wanted to verify: 
2 reasons behind not using Jenkins API and instead stick to GitHub API:
1. Jenkins API can be a bit restrictive. For e.g. There's a round-about way to get names of all the jobs ["checks" equivalent] corresponding to a branch [PR]. 
2. Instead of querying Jenkins server on AWS instance for details, we can query GitHub servers for same info instead.

Are there other compelling reasons to use GitHub API over Jenkins API? Just asking for my knowledge.

Comment (marcoabreu) [reply services/lambda-label-successful-prs/CIBot.py]: Checks API is the wrong API. You need the status API: https://docs.github.com/en/rest/reference/repos#statuses

Yes, Jenkins API can be a bit restrictive - also you won't cover other checks (e.g. Travis). GitHub already has the information, so there's no need to ask the source.

Yes, you can query the GitHub servers instead.

Biggest reason is that the GitHub API is the actual source of truth for our users. It doesn't matter if something in Jenkins is green or not - if that information is not present in GitHub, you can't merge. So if you forgot to add a Jenkins check or aren't querying Travis, they might still not be able to merge. The GitHub API knows which checks are required, which ones are present etc. Jenkins on the other hand does not.

Comment (marcoabreu) [reply services/lambda-label-successful-prs/CIBot.py]: Btw, you want a webhook based on the following event: https://developer.github.com/webhooks/event-payloads/#status

File: services/lambda-label-successful-prs/CIBot.py
Comment (ChaiBapchya) [reply services/lambda-label-successful-prs/CIBot.py]: Ya i hardcoded it. But with the github API we don't have to do this.

File: services/lambda-label-successful-prs/CIBot.py
Comment (ChaiBapchya) [reply services/lambda-label-successful-prs/CIBot.py]: If we are to use the webhook approach, can we request Apache Infra team to update the Events parameter to additionally send Statuses. 

Our previous request to the Infra team had following events:
```
Events : Issue Comments, Pull Requests, Pushes 
```
Refer: https://issues.apache.org/jira/browse/INFRA-20005

Can you help me with that request @marcoabreu Thanks.

Comment (marcoabreu) [reply services/lambda-label-successful-prs/CIBot.py]: Why not simply add a new webhook to keep the things separate?

Happy to assist.

Comment (marcoabreu) [reply services/lambda-label-successful-prs/CIBot.py]: Please don't forget to write a proposal on dev@ which describes what you're trying to do, what impact it's gonna have and what you need.

Comment (ChaiBapchya) [reply services/lambda-label-successful-prs/CIBot.py]: Do we really need a separate webhook when the 2 webhooks have everything in common except 1 event?

Previous webhook
```
URL : <>
Content type: application/json
Secret : <>
Events : Issue Comments, Pull Requests, Pushes
Enable SSL verification
```
New webhook
```
URL : <same url>
Content type: application/json
Secret : <>
Events : Issue Comments, Pull Requests, Pushes, Statuses
Enable SSL verification
```

I'm happy to write a proposal mentioning the usecase + impact + requesting addition of "statuses" in the existing webhook. Does that work with you?

Comment (marcoabreu) [reply services/lambda-label-successful-prs/CIBot.py]: Well if one thing breaks, do you want everything else to immediately break as well? It's just separation of concerns to reduce the blast radius and reduce the complexity of the application.

One bot is made for reacting to the comments.

This bot is made for reacting to status events.

File: services/lambda-label-successful-prs/CIBot.py
Comment (szha) [inline services/lambda-label-successful-prs/CIBot.py]: I think this will need to also take into account whether a PR is reviewed or not. For example, when I'm done reviewing a PR with "pr-awaiting-review", I may request change and remove that label, or change it to "pr-awaiting-response".

Comment (marcoabreu) [reply services/lambda-label-successful-prs/CIBot.py]: Yeah I'd certainly refrain from using the "pr-" labels but introducing new "ci-" labels. The former are manual labels while the latter are automated ones - don't mix them.

Comment (szha) [reply services/lambda-label-successful-prs/CIBot.py]: as long as the state machine is properly set up, it shouldn't matter whether a label was manually changed or not. Overall, I think labeling PRs with a status should be a fully automated

**Comment by szha** [comment]
maybe we could automate the PR labeling as the following pseudo code describes:
```
if WIP in title or PR is draft or CI failed:
    pr-work-in-progress
elif CI has not started yet or CI is in progress:
    pr-awaiting-testing
else: # CI passed checks
    if pr has at least one approval and no request changes:
        pr-awaiting-merge
    elif pr has no review or all reviews have been dismissed/re-requested:
        pr-awaiting-review
    else: # pr has a review that hasn't been dismissed yet no approval
        pr-awaiting-response
```

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (marcoabreu) [inline services/lambda-pr-status-labeler/PRStatusBot.py]: You can directly check if the label is inside the collection. No need to do that loop

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (marcoabreu) [inline services/lambda-pr-status-labeler/PRStatusBot.py]: Turn labels into constants at the top

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (marcoabreu) [inline services/lambda-pr-status-labeler/PRStatusBot.py]: Does this only count from committers?

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (marcoabreu) [inline services/lambda-pr-status-labeler/PRStatusBot.py]: I'd like to see some units tests for all these cases. Seems too complex for me, hence please add tests

**Review by marcoabreu — COMMENTED** [review]
Generally going in a great direction!

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: Labels is a PaginatedList of objects of type Label. Each object has an attribute `name`
I guess Python doesn't allow to check if objects in a list contain an attribute value without iteration.
I could use 
```
any(label.name == desired_label for label in labels)
```
But in general I'd have to iterate. Correct me if wrong.

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: No.. it counts for everyone.. should we only consider from committers?

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: Moved all labels/states as constants.

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (marcoabreu) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: Calling list(...) On it should iterate the entire paginated list automatically. Then, that operation should work. But I'll leave it up to you :)

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (marcoabreu) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: Yes, only committer reviews are "valid" reviews. Notice that a committer approval results in a green check mark while a non committer one is a grey check mark.

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: Sure. I'll leave it as this.

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: True. Missed that point. Added the check if reviewer is MXNet committer or not.

File: services/lambda-pr-status-labeler/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/PRStatusBot.py]: Added unit-tests.

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:277]: what about draft PR?

File: services/lambda-pr-status-labeler/pr_status_bot/test.py
Comment (marcoabreu) [inline services/lambda-pr-status-labeler/pr_status_bot/test.py:27]: Does this test against the real repo? I'm not that big of a fan of hitting servers when executing unit tests. They should be self enclosed.

File: services/lambda-pr-status-labeler/pr_status_bot/test.py
Comment (marcoabreu) [inline services/lambda-pr-status-labeler/pr_status_bot/test.py:51]: You are just executing the bot, but you are not actually validating anything besides if it throws an exception.

**Review by marcoabreu — CHANGES_REQUESTED** [review]
I think the approach to test should be reengineered. At the moment, this does not seem like a proper test procedure since things are not mocked out and neither are the results verified.

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:277]: draft PR is tested 2 lines below this.

**Comment by sandeep-krishnamurthy** [comment]
Not valid review comments. 
I just came here to thank you for the hard work on this feature and burning your weekend in keeping this feature developed.

**Comment by marcoabreu** [comment]
What do you mean with "not valid review comments", Sandeep?

**Comment by sandeep-krishnamurthy** [comment]
I just meant - My comment is not a technical review comment on the PR. This is just a thank you note :)

File: services/lambda-pr-status-labeler/pr_status_bot/test.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/test.py:27]: Yes it's hitting the real repo.

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:99]: move to constructor

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py]: move to constructor

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:177]: would you be able to always assign the appropriate label instead? this will make the operation idempotent.

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py]: extract to a method

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py]: is this really necessary for a method?

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:196]: `return desired_label in set(pr.get_labels())`

Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:196]: labels is a paginatedlist of Label objects [not strings]. Hence `set(pr.get_labels())` gives
```
TypeError: unhashable type: 'Label'
```

Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:196]: Hence I have to iterate over the list & retrieve the object attribute

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py]: dropped!

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py]: Factored out!

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:99]: `get_secret` is called inside constructor based on `apply_secret` parameter

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py]: Done!

File: services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/PRStatusBot.py:177]: Yes. assigning labels is idempotent. So I can drop the check

**Comment by ChaiBapchya** [comment]
To run all the tests
```
python -m pytest test_mock.py
```

File: services/lambda-pr-status-labeler/pr_status_bot/test_mock.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/test_mock.py:112]: use the real classes https://pygithub.readthedocs.io/en/latest/reference.html

File: services/lambda-pr-status-labeler/pr_status_bot/test.py
Comment (szha) [reply services/lambda-pr-status-labeler/pr_status_bot/test.py:51]: +1 this still needs to be addressed.

File: services/lambda-pr-status-labeler/pr_status_bot/test_mock.py
Comment (szha) [inline services/lambda-pr-status-labeler/pr_status_bot/test_mock.py]: use the real classes https://pygithub.readthedocs.io/en/latest/reference.html

**Review by szha — COMMENTED** [review]
remember to add apache header to all files.

File: services/lambda-pr-status-labeler/pr_status_bot/test_mock.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/test_mock.py]: Addressed

File: services/lambda-pr-status-labeler/pr_status_bot/test.py
Comment (ChaiBapchya) [reply services/lambda-pr-status-labeler/pr_status_bot/test.py:51]: Addressed in the script test: https://github.com/apache/incubator-mxnet-ci/blob/6131b9df5b1528661ec1991da90b3704c91f3c74/services/lambda-pr-status-labeler/pr_status_bot/test_mock.py#L64-L70

**Review by marcoabreu — APPROVED** [review]
LGTM, thanks for your work!

**Pull request merged**
